File: app/src/main/java/com/magic/upcoming/games/network/rx/BaseResponseFunc1.java
package com.magic.upcoming.games.network.rx;

import android.util.Log;

import com.magic.upcoming.games.network.exception.ResponseFailedException;
import com.magic.upcoming.games.network.response.BaseResponse;
import com.magic.upcoming.games.network.response.ErrorCode;
import com.magic.upcoming.games.network.response.ResponseCode;
import com.magic.upcoming.games.network.response.RxVoid;

import io.reactivex.Flowable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

/**
 * Created by shendawei on 17/2/8.
 */

public class BaseResponseFunc1<T> implements Function <BaseResponse <T>, Flowable <T>> {

    @Override
    public Flowable <T> apply(@NonNull final BaseResponse<T> response) throws Exception {

        if (response == null) {
            return Flowable.error(new ResponseFailedException(ErrorCode.NETWORK_ERROR,"网络出现问题1"));
        }
        Log.d("BaseResponse", "response = " + response);

        //如果请求操作成功且发射数据为空，则发送rxvoid
//        if(response.getStatus()==ResponseCode.SUCCESS&&response.getContent()==null)
//        {
//            return Flowable.just((T) new RxVoid());
//        }
        if(response.getContent()!=null)
        {
            return Flowable.just(response.getContent());
        }
        return Flowable.error(new ResponseFailedException(ErrorCode.NETWORK_ERROR,"网络出现问题2"));

    }
}

File: app/src/main/java/com/magic/upcoming/games/network/rx/BaseResponseFunc1.kt
package com.magic.upcoming.games.network.rx

import com.magic.upcoming.games.model.base.BaseModel
import com.magic.upcoming.games.network.exception.ResponseFailedException
import com.magic.upcoming.games.network.response.BaseResponse
import io.reactivex.Flowable
import io.reactivex.functions.Function

class BaseResponseFunc1<T> : Function<BaseResponse<T>, Flowable<BaseModel<T>>> {
    @Throws(Exception::class)
    override fun apply(response: BaseResponse<T>): Flowable<BaseModel<T>> {
        return when(response.statusCode){
            100 -> Flowable.error(ResponseFailedException("Invalid API Key"))
            101 -> Flowable.error(ResponseFailedException("Object Not Found"))
            102 -> Flowable.error(ResponseFailedException("Error in URL Format"))
            103 -> Flowable.error(ResponseFailedException("'jsonp' format requires a 'json_callback' argument"))
            104 -> Flowable.error(ResponseFailedException("Filter Error"))
            105 -> Flowable.error(ResponseFailedException("Subscriber only video is for subscribers only"))
            1 ->{
                val baseModel = BaseModel<T>()
                baseModel.result = response.results
                baseModel.limit = response.limit
                baseModel.offset = response.offset
                baseModel.resultsNum = response.numberOfPageResults
                baseModel.totalResults = response.numberOfTotalResults
                Flowable.just(baseModel)
            }
            else -> Flowable.error(ResponseFailedException("Network Error"))
        }
    }
}